### PR TITLE
Strict RFC 9562 validation for `uuid()`

### PR DESCRIPTION
**Pull request by dahlia**
Closes https://github.com/dahlia/optique/issues/336

`uuid()` documents itself as validating standard UUID strings, but the previous implementation accepted any hexadecimal nibble in the version position (including `0` and `9` through `f`) and did not validate the variant bits at all. This made it possible for non-standard or not-yet-defined UUID layouts to pass validation silently.

This change makes `uuid()` enforce RFC 9562 compliance by default. The version digit must now be one of the currently standardized versions (1 through 8), and the variant nibble at position 19 must have the RFC 9562 layout (`10xx` in binary, i.e., hex digits `8`, `9`, `a`, or `b`). The nil UUID (`00000000-0000-0000-0000-000000000000`) and max UUID (`ffffffff-ffff-ffff-ffff-ffffffffffff`) are accepted as special standard values regardless of strict mode.

For callers who need to accept UUID-like values with future or non-standard version digits, the `strict: false` option restores the previous lenient behavior:

```typescript
uuid();                  // strict by default
uuid({ strict: true });  // same as default
uuid({ strict: false }); // accept any hex version/variant (previous behavior)
```

When `allowedVersions` is provided, it takes precedence over the strict default version set (1 through 8), but variant validation still applies in strict mode. Nil and max UUIDs pass even with `allowedVersions` since they are standard special values that do not follow ordinary version/variant rules.

A new `errors.invalidVariant` callback has been added for custom error messages when variant validation fails, following the same pattern as the existing `errors.invalidUuid` and `errors.disallowedVersion` callbacks.

Files changed: *packages/core/src/valueparser.ts*, *packages/core/src/valueparser.test.ts*, *docs/concepts/valueparsers.md*, *CHANGES.md*.

### CODE REVIEW

**Comment by dahlia** [comment]
@coderabbitai full review

File: docs/concepts/valueparsers.md [hunk 1]
```diff
@@ -507,10 +508,35 @@ The parser validates the standard UUID format:
...
 ~~~~
 
+### Strict RFC 9562 validation
+
+By default, the parser enforces strict RFC 9562 compliance:
+
+ -  The *version digit* (position 14) must be one of the currently
+    standardized versions: 1 through 8.
+ -  The *variant nibble* (position 19) must have the RFC 9562 layout
+    (`10xx` in binary), meaning hex digits `8`, `9`, `a`, or `b`.
```
[the remark this answers is not in the document]
Comment (dahlia) [reply docs/concepts/valueparsers.md]: Fixed in 5fccbbe. Changed to use 0-based index with a note on first mention: "index 14, 0-based" and "index 19".

File: docs/concepts/valueparsers.md [hunk 2]
```diff
@@ -507,10 +508,35 @@ The parser validates the standard UUID format:
...
+
+The *nil UUID* (`00000000-0000-0000-0000-000000000000`) and *max UUID*
+(`ffffffff-ffff-ffff-ffff-ffffffffffff`) are accepted as special standard
+values regardless of strict mode.
```
[the remark this answers is not in the document]
Comment (dahlia) [reply docs/concepts/valueparsers.md]: Fixed in 5fccbbe. The docs now say "regardless of strict mode or `allowedVersions`".

File: packages/core/src/valueparser.test.ts [hunk 3]
```diff
@@ -3896,6 +3896,254 @@ describe("uuid", () => {
...
+      ];
+
+      for (const input of cases) {
+        assert.equal(
+          defaultParser.parse(input).success,
+          strictParser.parse(input).success,
+          `Mismatch for: ${input}`,
+        );
+      }
+    });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.test.ts:3999]: Fixed in 5fccbbe. Changed to `assert.deepEqual` to compare the full parse result.

File: packages/core/src/valueparser.test.ts [hunk 4]
```diff
@@ -3896,6 +3896,254 @@ describe("uuid", () => {
...
+      const result = parser.parse("550e8400-e29b-41d4-0716-446655440000");
+      assert.ok(!result.success);
+    });
+
+    it("should skip variant check with allowedVersions and strict: false", () => {
+      const parser = uuid({ allowedVersions: [4], strict: false });
+      // v4 UUID with invalid variant nibble (0)
+      const result = parser.parse("550e8400-e29b-41d4-0716-446655440000");
+      assert.ok(result.success);
+    });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.test.ts:4037]: Fixed in 5fccbbe. Added a test case that verifies `allowedVersions` still rejects disallowed versions when `strict: false`.

File: packages/core/src/valueparser.ts [hunk 5]
```diff
@@ -1794,11 +1831,13 @@ export function uuid(options: UuidOptions = {}): ValueParser<"sync", Uuid> {
   const metavar = options.metavar ?? "UUID";
   ensureNonEmptyString(metavar);
   // Snapshot mutable config at construction time
+  const strict = options.strict !== false;
   const allowedVersions = options.allowedVersions != null
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts:1838]: Fixed in 5fccbbe. Added `checkBooleanOption(options, "strict")` before snapshotting.

File: packages/core/src/valueparser.ts [hunk 6]
```diff
@@ -1845,6 +1893,36 @@ export function uuid(options: UuidOptions = {}): ValueParser<"sync", Uuid> {
...
+      } else if (strict && (version < 1 || version > 8)) {
+        // In strict mode without allowedVersions, require RFC 9562 versions
+        return {
+          success: false,
+          error: disallowedVersion
+            ? (typeof disallowedVersion === "function"
+              ? disallowedVersion(version, [1, 2, 3, 4, 5, 6, 7, 8])
+              : disallowedVersion)
+            : message`Expected UUID version 1 through 8, but got version ${
+              version.toLocaleString("en")
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts:1908]: The existing `allowedVersions` code path already uses `v.toLocaleString("en")` for the same purpose. Since the error messages themselves are in English, using the English locale for number formatting is intentional and consistent here.

File: CHANGES.md [hunk 7]
```diff
@@ -616,6 +616,14 @@ To be released.
...
     [[#344], [#653]]
 
+ -  `uuid()` now enforces strict [RFC 9562] validation by default: the
+    version digit must be 1 through 8, and the variant nibble must follow
+    the RFC 9562 layout (`10xx`).  The nil and max UUIDs are accepted as
+    special standard values.  Use `uuid({ strict: false })` for the previous
+    lenient behavior that accepts any hex digit in the version and variant
+    positions.  [[#336], [#670]]
+
+[RFC 9562]: https://www.rfc-editor.org/rfc/rfc9562
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md:626]: This placement follows the Hongdown Markdown formatter rules used in this project. Hongdown places reference links immediately after the paragraph that first uses them, not at the end of the section.

File: docs/concepts/valueparsers.md [hunk 8]
```diff
@@ -507,10 +508,35 @@ The parser validates the standard UUID format:
 550e8400e29b41d4a716446655440000       # Missing hyphens
 ~~~~
 
+### Strict RFC 9562 validation
```
[the remark this answers is not in the document]
Comment (dahlia) [reply docs/concepts/valueparsers.md:511]: This is a subsection within the "uuid() parser" section, so ATX style (`###`) is correct per the project's Markdown style guide: "ATX-style headings: Use only for subsections within a section."

File: packages/core/src/valueparser.test.ts [hunk 9]
```diff
@@ -3896,6 +3896,254 @@ describe("uuid", () => {
...
+        "550e8400-e29b-41d4-0716-446655440000", // variant 0
+        "550e8400-e29b-41d4-f716-446655440000", // variant f
+        "01234567-89ab-cdef-0123-456789abcdef", // non-standard
+      ];
+
+      for (const input of cases) {
+        const result = parser.parse(input);
+        assert.ok(result.success, `Should accept with strict: false: ${input}`);
+      }
+    });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.test.ts:4016]: Fixed in d4c6dde. Added a test that asserts TypeError for non-boolean strict values (1, "true", 0).

File: packages/core/src/valueparser.ts [hunk 10]
```diff
@@ -1780,9 +1812,14 @@ export interface UuidOptions {
...
  * This parser validates that the input is a well-formed UUID string in the
  * standard format: `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` where each `x`
- * is a hexadecimal digit.  The parser can optionally restrict to specific
- * UUID versions.
+ * is a hexadecimal digit.
+ *
+ * By default, the parser enforces strict [RFC 9562] validation: it requires
+ * a standardized version digit (1 through 8) and the RFC 9562 variant bits
+ * (`10xx`).  The nil and max UUIDs are accepted as special standard values.
+ * Set `strict: false` to disable version and variant validation.
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts]: Fixed in d4c6dde. The JSDoc now reads: "Set `strict: false` to disable the default RFC 9562 version/variant checks. An explicit `allowedVersions` list still constrains the version nibble even in lenient mode."

File: packages/core/src/valueparser.test.ts [hunk 11]
```diff
@@ -3896,6 +3896,261 @@ describe("uuid", () => {
...
+      const parser = uuid({ allowedVersions: [4] });
+      const result = parser.parse("00000000-0000-0000-0000-000000000000");
+      assert.ok(result.success);
+    });
+
+    it("should accept max UUID even with allowedVersions", () => {
+      const parser = uuid({ allowedVersions: [4] });
+      const result = parser.parse("ffffffff-ffff-ffff-ffff-ffffffffffff");
+      assert.ok(result.success);
+    });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.test.ts:4049]: Fixed in c89574c. Added a test for nil and max UUIDs with `allowedVersions: [4], strict: false`.

File: packages/core/src/valueparser.ts [hunk 12]
```diff
@@ -1748,10 +1748,35 @@ export interface UuidOptions {
...
+   * the {@link strict} option.
    */
   readonly allowedVersions?: readonly number[];
 
+  /**
+   * Whether to enforce strict [RFC 9562] validation.  When `true` (the
+   * default), the parser validates that the version digit is one of the
+   * currently standardized versions (1 through 8) and that the variant bits
+   * follow the RFC 9562 layout (`10xx`, i.e., hex digits `8`, `9`, `a`,
+   * or `b` at position 20 of the UUID string).
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts:1761]: The error message uses "position 20" (1-based) intentionally because it is a user-facing string. Users counting characters in a UUID string will naturally use 1-based positions. The JSDoc and docs use 0-based "index 19" for developers. This is consistent with how other tools present character positions to end users.

File: packages/core/src/valueparser.test.ts [hunk 13]
```diff
@@ -3917,11 +4197,18 @@ describe("uuid", () => {
...
-      assert.ok(result.success); // Format is correct, even if version is unusual
+      assert.ok(!result.success);
+    });
+
+    it("should accept version 0 with strict: false", () => {
+      const parser = uuid({ strict: false });
+      const result = parser.parse("6ba7b800-9dad-01d1-80b4-00c04fd430c8"); // v0
+
+      assert.ok(result.success);
     });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.test.ts:4212]: These edge case tests were updated from the pre-existing "should handle version 0" test. The strict mode block tests the feature systematically, while the edge cases block retains the original boundary-value perspective. Keeping them separate preserves the existing test organization.
